crates/ty_python_semantic/resources/mdtest/import/nonstandard_conventions.md

```diff
@@ -1,39 +1,35 @@
-to encourage intentional API design. Although `.pyi` files are explicitly designed to work with
-typecheckers, which ostensibly should all enforce these strict rules, every typechecker has its own
-defacto "extensions" to them and so a few idioms like `from . import mysubmodule` have found their
-way into `.pyi` files too.
-
-Thus for the sake of compatibility, we need to define our own "extensions". Any extensions we define
-here have several competing concerns:
-
-- Extensions should ideally be kept narrow to continue to encourage explicit API design
-- Extensions should be easy to explain, document, and understand
-- Extensions should ideally still be a subset of runtime behaviour (if it works in a stub, it works
-    at runtime)
-- Extensions should ideally not make `.pyi` files more permissive than `.py` files (if it works in a
-    stub, it works in an impl)
-
-To that end we define the following extension:
-
-> If an `__init__.pyi` for `mypackage` contains a `from...import` targetting a direct submodule of
-> `mypackage`, then that submodule should be available as an attribute of `mypackage`.
+[standard import conventions](https://typing.python.org/en/latest/spec/distributing.html#import-conventions),
+and other intentional deviations from actual python semantics.
+
+This file currently covers the following details:
+
+- **froms are locals**: a `from..import` can only define locals, it does not have global
+    side-effects. Specifically any submodule attribute `a` that's implicitly introduced by either
+    `from .a import b` or `from . import a as b` (in an `__init__.py(i)`) is a local and not a
+    global. If you do such an import at the top of a file you won't notice this. However if you do
+    such an import in a function, that means it will only be function-scoped (so you'll need to do
+    it in every function that wants to access it, making your code less sensitive to execution
+    order).
+
+- **first from first serve**: only the *first* `from..import` in an `__init__.py(i)` that imports a
+    particular direct submodule of the current package introduces that submodule as a local.
+    Subsequent imports of the submodule will not introduce that local. This reflects the fact that
+    in actual python only the first import of a submodule (in the entire execution of the program)
+    introduces it as an attribute of the package. By "first" we mean "the first time in this scope
+    (or any parent scope)". This pairs well with the fact that we are specifically introducing a
+    local (as long as you don't accidentally shadow or overwrite the local).
+
+- **dot re-exports**: `from . import a` in an `__init__.pyi` is considered a re-export of `a`
+    (equivalent to `from . import a as a`). This is required to properly handle many stubs in the
+    wild. Currently it must be *exactly* `from . import ...`.
 
 ## Relative `from` Import of Direct Submodule in `__init__`
```

Comment:
Nit: the fact that all test headers / names in this test file assume "in a stub file" by default (and then add "(Non-stub Check)" to the equivalent non-stub test) may have made sense before this PR, when this file was only about modifying stub file rules, but it feels weird after this PR. I would probably put the non-stub test first in each scenario, and explicitly mark the stub tests instead? (Or really I think I would just have each test header once and then use sub-headings `### In py file` and `### In stub file` rather than repeating the whole test description.

Reply:
You're extremely right and I was hoping no one would notice. *rolls up sleeves*